### PR TITLE
Ensure that on first call to ReadService.stateUpdates beginOffset is …

### LINKED ISSUE #6023: First call to ReadService.stateUpdates should have an unset beginOffset
When a ReadService is called for the first time the beginOffset passed in the stateUpdates call should be None, not Some(Offset.begin).

Subsequent calls should continue to use the offset from the last processed update.

### PR DESCRIPTION
**Pull request by simonmaxen-da**
…None

Fixes #6023

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala [hunk 1]
```diff
@@ -110,11 +110,15 @@ private class JdbcLedgerDao(
...
+  /**
+    * Defaults to Offset.begin if ledger_end is unset
+    */
+  override def lookupLedgerEnd(): Future[Offset] = {
     dbDispatcher.executeSql(metrics.daml.index.db.getLedgerEnd) { implicit conn =>
       SQL_SELECT_LEDGER_END
-        .as(offset("ledger_end").single)
+        .as(offset("ledger_end").?.map(_.getOrElse(Offset.begin)).single)
     }
+  }
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:121]: The braces are unnecessary.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala [hunk 2]
```diff
@@ -110,11 +110,15 @@ private class JdbcLedgerDao(
...
 
-  override def lookupLedgerEnd(): Future[Offset] =
+  /**
+    * Defaults to Offset.begin if ledger_end is unset
+    */
+  override def lookupLedgerEnd(): Future[Offset] = {
     dbDispatcher.executeSql(metrics.daml.index.db.getLedgerEnd) { implicit conn =>
       SQL_SELECT_LEDGER_END
-        .as(offset("ledger_end").single)
+        .as(offset("ledger_end").?.map(_.getOrElse(Offset.begin)).single)
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:119]: I think that when this function is called, there should _always_ be a value here, so crashing would be appropriate. If we haven't initialized the ledger, the code should be calling `lookupInitialLedgerEnd` instead.

What happens if you revert this? What breaks?

**Comment by SamirTalwar-DA** [comment]
Fixes #6023.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala (hunk 2 above)
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:119]: Yes, that's correct, @SamirTalwar-DA.

Comment (simonmaxen-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:119]: The previous behaviour was:
1. If called prior to ledger initialisation - _crash_
2. If called post initialisation but pre first ledger end update - _default to Offset.begin_
3. If called post first ledger end update - _return most recently updated ledger_end_

This change means that the default provided in #2 above comes from the method as opposed to from the database column.

As many things could be happening asynchronously during startup I think it is safest to leave the default, or leave it someone more expert in the area to remove it ;-)

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala (hunk 1 above)
Comment (simonmaxen-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala]: Fixed.

**Comment by simonmaxen-da** [comment]
@SamirTalwar-DA - I've updated the comment wrt 'Fixes'.  More importantly I've reverted some poor advice I previously gave in CONTRIBUTING.md on this subject.

**Comment by SamirTalwar-DA** [comment]
You need to update the PR description, not the commit message. 😉

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -21,7 +21,7 @@ For Git commit messages, our principle is that `git log --pretty=oneline` should
   * *Improve explanation of …*
   * *Remove module X because it is not used.*
 * Commits should have a description that concisely explains the rationale and context for the change if that is not obvious.
-* The first line of the description should include a reference to the fixed issue `(#XXX)`.
+* Commit descriptions should include a `Fixes #XX` line indicating what GitHub issue number the commit fixes.
```
Comment (SamirTalwar-DA) [inline CONTRIBUTING.md:24]:
```suggestion
* Commit descriptions can include a `Fixes #XX` line indicating what GitHub issue number the commit fixes.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -35,14 +35,14 @@ For Git commit messages, our principle is that `git log --pretty=oneline` should
 
 The following is an example of a well-formed commit, including the description (first line) and a body that includes changelog additions:
 
-      Introduced a new API for package management (#1311)
+      Fixes #1311
```
Comment (SamirTalwar-DA) [inline CONTRIBUTING.md:38]: That's not a useful first line. The previous one was better.

We tend to use the present tense in commit messages, so I'd suggest "Introduces" or "Introduce" over "Introduced".

```suggestion
      Introduces a new API for package management

      Fixes #1311.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -35,14 +35,14 @@ For Git commit messages, our principle is that `git log --pretty=oneline` should
...
-      Also fixes a typo in the Scala bindings documentation
+      Also fixes a typo in the Scala bindings documentation.
 
       CHANGELOG_BEGIN
 
       - [Sandbox] Introduced a new API for package management.
-      See `#1311 <https://github.com/digital-asset/daml/issues/1311>`__.
+        See `#1311 <https://github.com/digital-asset/daml/issues/1311>`__.
 
       CHANGELOG_END
```
Comment (SamirTalwar-DA) [inline CONTRIBUTING.md:47]: I have no idea why this document uses 6 spaces for indenting code, but it really confuses GitHub. I suggest changing it to use 4 spaces.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala [hunk 6]
```diff
@@ -139,7 +141,7 @@ private class JdbcLedgerDao(
   // and thus we need to make sure to only update the ledger end when the ledger entry we're committing
   // is advancing it.
   private val SQL_UPDATE_LEDGER_END = SQL(
-    "update parameters set ledger_end = {LedgerEnd} where ledger_end < {LedgerEnd}")
+    "update parameters set ledger_end = {LedgerEnd} where (ledger_end is null or ledger_end < {LedgerEnd})")
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:144]:
```suggestion
    "update parameters set ledger_end = {LedgerEnd} where ledger_end is null or ledger_end < {LedgerEnd}")
```

File: CONTRIBUTING.md (hunk 3 above)
Comment (simonmaxen-da) [reply CONTRIBUTING.md:24]: As discussed with Samir this was a revert of CONTRIBUTING.md.

File: CONTRIBUTING.md (hunk 4 above)
Comment (simonmaxen-da) [reply CONTRIBUTING.md:38]: as above

File: CONTRIBUTING.md (hunk 5 above)
Comment (simonmaxen-da) [reply CONTRIBUTING.md:47]: as above

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala (hunk 6 above)
Comment (simonmaxen-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala:144]: as above

**Pull request merged**